File: iOS/Objc/Examples/cases/BucketTagging.m
#import <XCTest/XCTest.h>
#import <QCloudCOSXML/QCloudCOSXML.h>
#import <QCloudCOSXML/QCloudUploadPartRequest.h>
#import <QCloudCOSXML/QCloudCompleteMultipartUploadRequest.h>
#import <QCloudCOSXML/QCloudAbortMultipfartUploadRequest.h>
#import <QCloudCOSXML/QCloudMultipartInfo.h>
#import <QCloudCOSXML/QCloudCompleteMultipartUploadInfo.h>


@interface BucketTagging : XCTestCase <QCloudSignatureProvider, QCloudCredentailFenceQueueDelegate>

@property (nonatomic) QCloudCredentailFenceQueue* credentialFenceQueue;

@end

@implementation BucketTagging

- (void)setUp {
    // 注册默认的 COS 服务
    QCloudServiceConfiguration* configuration = [QCloudServiceConfiguration new];
    configuration.appID = @"1250000000";
    configuration.signatureProvider = self;
    QCloudCOSXMLEndPoint* endpoint = [[QCloudCOSXMLEndPoint alloc] init];
    endpoint.regionName = @"ap-guangzhou";//服务地域名称，可用的地域请参考注释
    configuration.endpoint = endpoint;
    [QCloudCOSXMLService registerDefaultCOSXMLWithConfiguration:configuration];
    [QCloudCOSTransferMangerService registerDefaultCOSTransferMangerWithConfiguration:configuration];
    
    // 脚手架用于获取临时密钥
    self.credentialFenceQueue = [QCloudCredentailFenceQueue new];
    self.credentialFenceQueue.delegate = self;
}

- (void) fenceQueue:(QCloudCredentailFenceQueue * )queue requestCreatorWithContinue:(QCloudCredentailFenceQueueContinue)continueBlock
{
    QCloudCredential* credential = [QCloudCredential new];
    //在这里可以同步过程从服务器获取临时签名需要的 secretID，secretKey，expiretionDate 和 token 参数
    credential.secretID = @"COS_SECRETID";
    credential.secretKey = @"COS_SECRETKEY";
    credential.token = @"COS_TOKEN";
    /*强烈建议返回服务器时间作为签名的开始时间，用来避免由于用户手机本地时间偏差过大导致的签名不正确 */
    credential.startDate = [[[NSDateFormatter alloc] init] dateFromString:@"startTime"]; // 单位是秒
    credential.experationDate = [[[NSDateFormatter alloc] init] dateFromString:@"expiredTime"];
    QCloudAuthentationV5Creator* creator = [[QCloudAuthentationV5Creator alloc]
                                            initWithCredential:credential];
    continueBlock(creator, nil);
}

- (void) signatureWithFields:(QCloudSignatureFields*)fileds
                     request:(QCloudBizHTTPRequest*)request
                  urlRequest:(NSMutableURLRequest*)urlRequst
                   compelete:(QCloudHTTPAuthentationContinueBlock)continueBlock
{
    [self.credentialFenceQueue performAction:^(QCloudAuthentationCreator *creator,
                                               NSError *error) {
        if (error) {
            continueBlock(nil, error);
        } else {
            QCloudSignature* signature =  [creator signatureForData:urlRequst];
            continueBlock(signature, nil);
        }
    }];
}

/**
 * 用于为存储桶设置键值对作为存储桶标签，可以协助您管理已有的存储桶资源，并通过标签进行成本管理。
 */
- (void)putBucketTagging {
    
    //.cssg-snippet-body-start:[objc-put-bucket-tagging]
    QCloudPutBucketTaggingRequest *putReq = [QCloudPutBucketTaggingRequest new];
    
    // 存储桶名称，格式为 BucketName-APPID
    putReq.bucket = @"examplebucket-1250000000";
    
    // 标签集合
    QCloudBucketTagging *taggings = [QCloudBucketTagging new];
    
    QCloudBucketTag *tag1 = [QCloudBucketTag new];
    
    // 标签的 Key，长度不超过128字节, 支持英文字母、数字、空格、加号、减号、下划线、等号、点号、
    // 冒号、斜线
    tag1.key = @"age";
    
    // 标签的 Value，长度不超过256字节, 支持英文字母、数字、空格、加号、减号、下划线、等号、点号
    // 、冒号、斜线
    tag1.value = @"20";
    QCloudBucketTag *tag2 = [QCloudBucketTag new];
    tag2.key = @"name";
    tag2.value = @"karis";
    
    // 标签集合，最多支持10个标签
    QCloudBucketTagSet *tagSet = [QCloudBucketTagSet new];
    tagSet.tag = @[tag1,tag2];
    taggings.tagSet = tagSet;
    
    // 标签集合
    putReq.taggings = taggings;
    
    [putReq setFinishBlock:^(id outputObject, NSError *error) {
        // outputObject 包含所有的响应 http 头部
        NSDictionary* info = (NSDictionary *) outputObject;
    }];
    [[QCloudCOSXMLService defaultCOSXML] PutBucketTagging:putReq];
    
    //.cssg-snippet-body-end
    
    
}

/**
 * 用于查询指定存储桶下已有的存储桶标签。
 */
- (void)getBucketTagging {
    
    //.cssg-snippet-body-start:[objc-get-bucket-tagging]
    QCloudGetBucketTaggingRequest *getReq = [QCloudGetBucketTaggingRequest new];
    
    // 存储桶名称，格式为 BucketName-APPID
    getReq.bucket = @"examplebucket-1250000000";
    
    [getReq setFinishBlock:^(QCloudBucketTagging * result, NSError * error) {
        // tag的集合
        QCloudBucketTagSet * tagSet = result.tagSet;
    }];
    [[QCloudCOSXMLService defaultCOSXML] GetBucketTagging:getReq];
    
    //.cssg-snippet-body-end
    
}

/**
 * 用于删除指定存储桶下已有的存储桶标签。
 */
- (void)deleteBucketTagging {
    
    
    //.cssg-snippet-body-start:[objc-delete-bucket-tagging]
    QCloudDeleteBucketTaggingRequest *delReq = [QCloudDeleteBucketTaggingRequest new];
    
    // 存储桶名称，格式为 BucketName-APPID
    delReq.bucket =  @"examplebucket-1250000000";
    
    [delReq setFinishBlock:^(id outputObject, NSError *error) {
        // outputObject 包含所有的响应 http 头部
        NSDictionary* info = (NSDictionary *) outputObject;
    }];
    [[QCloudCOSXMLService defaultCOSXML] DeleteBucketTagging:delReq];
    
    //.cssg-snippet-body-end
    
}
// .cssg-methods-pragma


- (void)testBucketTagging {
    // 设置存储桶标签
    [self putBucketTagging];
    
    // 获取存储桶标签
    [self getBucketTagging];
    
    // 删除存储桶标签
    [self deleteBucketTagging];
    // .cssg-methods-pragma
    
}

@end

File: iOS/Swift/Examples/cases/BucketTagging.swift
import XCTest
import QCloudCOSXML

class BucketTagging: XCTestCase,QCloudSignatureProvider,QCloudCredentailFenceQueueDelegate{
    
    var credentialFenceQueue:QCloudCredentailFenceQueue?;
    
    override func setUp() {
        let config = QCloudServiceConfiguration.init();
        config.signatureProvider = self;
        config.appID = "1253653367";
        let endpoint = QCloudCOSXMLEndPoint.init();
        endpoint.regionName = "ap-guangzhou";//服务地域名称，可用的地域请参考注释
        endpoint.useHTTPS = true;
        config.endpoint = endpoint;
        QCloudCOSXMLService.registerDefaultCOSXML(with: config);
        QCloudCOSTransferMangerService.registerDefaultCOSTransferManger(with: config);
        
        // 脚手架用于获取临时密钥
        self.credentialFenceQueue = QCloudCredentailFenceQueue();
        self.credentialFenceQueue?.delegate = self;
    }
    
    func fenceQueue(_ queue: QCloudCredentailFenceQueue!,
                    requestCreatorWithContinue continueBlock: QCloudCredentailFenceQueueContinue!) {
        let cre = QCloudCredential.init();
        //在这里可以同步过程从服务器获取临时签名需要的 secretID，secretKey，expiretionDate 和 token 参数
        cre.secretID = "COS_SECRETID";
        cre.secretKey = "COS_SECRETKEY";
        cre.token = "COS_TOKEN";
        /*强烈建议返回服务器时间作为签名的开始时间，用来避免由于用户手机本地时间偏差过大导致的签名不正确 */
        cre.startDate = DateFormatter().date(from: "startTime"); // 单位是秒
        cre.experationDate = DateFormatter().date(from: "expiredTime");
        let auth = QCloudAuthentationV5Creator.init(credential: cre);
        continueBlock(auth,nil);
    }
    
    func signature(with fileds: QCloudSignatureFields!,
                   request: QCloudBizHTTPRequest!,
                   urlRequest urlRequst: NSMutableURLRequest!,
                   compelete continueBlock: QCloudHTTPAuthentationContinueBlock!) {
        self.credentialFenceQueue?.performAction({ (creator, error) in
            if error != nil {
                continueBlock(nil,error!);
            }else{
                let signature = creator?.signature(forData: urlRequst);
                continueBlock(signature,nil);
            }
        })
    }
    
    
    /**
     * 用于为存储桶设置键值对作为存储桶标签，可以协助您管理已有的存储桶资源，并通过标签进行成本管理。
     */
    func putBucketTagging() {
        //.cssg-snippet-body-start:[swift-put-bucket-tagging]
        let req = QCloudPutBucketTaggingRequest.init();
        
        // 存储桶名称，格式为 BucketName-APPID
        req.bucket = "examplebucket-1250000000";
        let taggings = QCloudBucketTagging.init();
        
        // 标签集合
        let tagSet = QCloudBucketTagSet.init();
        taggings.tagSet = tagSet;
        let tag1 = QCloudBucketTag.init();
        
        // 标签的 Key，长度不超过128字节, 支持英文字母、数字、空格、加号、减号、下划线、等号、点号、
        // 冒号、斜线
        tag1.key = "age";
        
        // 标签的 Value，长度不超过256字节, 支持英文字母、数字、空格、加号、减号、下划线、等号、点号
        // 、冒号、斜线
        tag1.value = "20";
        
        let tag2 = QCloudBucketTag.init();
        tag2.key = "name";
        tag2.value = "karis";
        
        // 标签集合，最多支持10个标签
        tagSet.tag = [tag1,tag2];
        
        // 标签集合
        req.taggings = taggings;
        req.finishBlock = {(result,error) in
            if let result = result {
                // result 包含响应的 header 信息
            } else {
                print(error!);
            }
        }
        QCloudCOSXMLService.defaultCOSXML().putBucketTagging(req);
        
        //.cssg-snippet-body-end
    }
    
    
    /**
     * 用于查询指定存储桶下已有的存储桶标签。
     */
    func getBucketTagging() {
        //.cssg-snippet-body-start:[swift-get-bucket-tagging]
        let req = QCloudGetBucketTaggingRequest.init();
        
        // 存储桶名称，格式为 BucketName-APPID
        req.bucket = "examplebucket-1250000000";
        req.setFinish { (result, error) in
            if let result = result {
                // 标签集合
                let tagSet = result.tagSet
            } else {
                print(error!);
            }
        };
        QCloudCOSXMLService.defaultCOSXML().getBucketTagging(req);
        
        //.cssg-snippet-body-end
    }
    
    
    /**
     * 用于删除指定存储桶下已有的存储桶标签。
     */
    func deleteBucketTagging() {
        //.cssg-snippet-body-start:[swift-delete-bucket-tagging]
        let req = QCloudDeleteBucketTaggingRequest.init();
        
        // 存储桶名称，格式为 BucketName-APPID
        req.bucket = "examplebucket-1250000000";
        req.finishBlock =  { (result, error) in
            if let result = result {
                // result 包含响应的 header 信息
            } else {
                print(error!);
            }
        };
        QCloudCOSXMLService.defaultCOSXML().deleteBucketTagging(req);
        
        //.cssg-snippet-body-end
    }
    // .cssg-methods-pragma
    
    func testBucketTagging() {
        // 设置存储桶标签
        self.putBucketTagging();
        // 获取存储桶标签
        self.getBucketTagging();
        // 删除存储桶标签
        self.deleteBucketTagging();
        // .cssg-methods-pragma
    }
}
